### PR TITLE
Update cice.t-test.py to use cartopy instead of basemap.

### PR DESCRIPTION
**Pull request by daveh150**
For detailed information about submitting Pull Requests (PRs) to the CICE-Consortium,
please refer to: <https://github.com/CICE-Consortium/About-Us/wiki/Resource-Index#information-for-developers>

## PR checklist
- [x] Short (1 sentence) summary of your PR: 
Update cice.t-test.py to use cartopy instead of basemap.

- [x] Developer(s): 
    David Hebert
- [x] Suggest PR reviewers from list in the column to the right.
- [ ] Please copy the PR test results link or provide a summary of testing completed below.
     
- How much do the PR code changes differ from the unmodified code? 
    - [ x] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [x ] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [ x] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [ x] Yes
    - [ ] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [ x] No 
- [ ] Please provide any additional information or relevant details below:

### CODE REVIEW

**Comment by eclare108213** [comment]
@daveh150
These are fairly extensive changes -- it took some work to do this.  Thanks!  Do you have sample output, before and/or at least after?

For future reference (and you can edit your text above) - deleting the space from the checkbox code when you add an x will make them work...   [x] instead of [ x] etc.

**Comment by daveh150** [comment]
Thank you. The contour plot section was the challenging part, needing both +/- 180 and 0-360 lat/lons.

I admittedly did not test the plot in this exact script, I do not have a perturbation run to test with (I likely missed how to set that up on the testing documentation).  I was testing plots outside of this script using output from one the 'smoke' tests.  If anyone has suggestion on how to make a perturbation test I can test this.

**Comment by daveh150** [comment]
Oh, I think I see why I couldn't run a QC test before. Previously I got a message that my test and baseline were bit-for-bit, so no qc test was required. But I did not make any changes to the ice_in, so the runs were identical. I'll try to change the shortwave and ktherm in the 'test' run and see if I can run the QC test to get example plots from this script. Sorry about that.

**Comment by daveh150** [comment]
Does anyone use the contour option in these plots? That is by far the most challenging, the scatter and pcolor seem to do what we want. I'd like to just remove the contour plots if they aren't used.

**Comment by eclare108213** [comment]
I didn't know that contour plotting was even available.  As far as I'm concerned, color maps should be sufficient and you can comment the contouring part out.  (Am I correctly understanding the issue here?) Check the docs to see if it's mentioned, and if so then comment that out too.  Thx

**Comment by apcraig** [comment]
This looks fine to me, I'll try to run a t-test on cheyenne and make sure everything is working to independently confirm.  I've struggled at times getting the python environment setup, has anything changed in that realm?

**Comment by apcraig** [comment]
Thanks @daveh150, those plots look pretty good overall.  No idea about the missing data in the contour plots.

**Comment by daveh150** [comment]
I made an update where if contour is selected, the plots will revert to pcolor. Eventually that whole section of the code could be removed once we are sure no one wants to use it. It cleans up the plotting code significantly. I did not see a need to add the cyclic arrays with the pcolormesh.

**Comment by dabail10** [comment]
Just got a chance to look at this. This is great! Thanks @daveh150 for doing this. Here are the plots from my gx3 B vs C grid comparison. A couple of notes for enhancements later.

1. This does seem to be in the right sense of the differences I would expect. That is the B grid solution is thinner.
2. However, It would be nice if the colorbar could be reset a bit to bring out the differences. The "red" area is basically zero everywhere and so a red-white-blue colormap might be nice for the difference plot. Then instead of having the maximum distance as the colorbar, maybe a one-sigma difference is better. That is computing the spatial variance of the distances and using that as the max on the colorbar instead of the absolute max difference. Perhaps even two-sigma. This would help bring out the smaller differences better.
3. Also, it would be nice if the colorbar for the two cases was the same. This way one could see the differences pretty clearly in these plots.

ice_thickness_gx3testb
ice_thickness_gx3testc
ice_thickness_gx3testb_minus_gx3testc

**Review by dabail10 — APPROVED** [review]
This looks great. I added some additional comments for enhancements.

**Comment by dabail10** [comment]
> I made an update where if contour is selected, the plots will revert to pcolor. Eventually that whole section of the code could be removed once we are sure no one wants to use it. It cleans up the plotting code significantly. I did not see a need to add the cyclic arrays with the pcolormesh.

I agree that it is probably not worth chasing down the issues with the contour plots. Typically we don't see issues like this except on tripolar grids. It also looks like the line of missing values goes along the 90E/W line. It would make more sense if it was related to 180/0. I recall that Matt had designed it so that one could choose in case the pcolor doesn't work for some reason. What happens if you don't use filled contours?

**Comment by apcraig** [comment]
@dabail10, thanks for reviewing and testing the scripts.  @daveh150, Are we ready to merge?  It seems like the outstanding requests may be

- update environment.yml, add cartopy, remove basemap
- update the colorbar

I think those are two good suggestions.  If we prefer to defer, I'm also OK merging.

**Comment by apcraig** [comment]
@daveh150, just checking in again to see how we should proceed.

**Comment by apcraig** [comment]
Thanks @daveh150.  @dabail10, do you have some QC data you can run the updated scripts on to verify they work and look good?  If not, I will try to generate some data and do the same.  thanks!

**Comment by dabail10** [comment]
I do have some runs I can test this on.

**Comment by dabail10** [comment]
These are great! It is great to see such a large area of white where the differences are close to zero. We might even be able to set the min/max on the difference plots to even smaller, but this is great for now.

ice_thickness_gx1testb_minus_gx1testc
ice_thickness_gx1testc
ice_thickness_gx1testb

**Comment by daveh150** [comment]
@dabail10 , interesting, those plots are not what I was expecting. Do you have the latest version of the script? I set it up to have colorbars and colormaps like the attached. Maybe I didn't commit something correctly?

ice_thickness_nrlssc_gnu_smoke_gx3_20x1_qc qc_base

ice_thickness_nrlssc_gnu_smoke_gx3_20x1_qc qc_te

ice_thickness_nrlssc_gnu_smoke_gx3_20x1_qc qc_base_minus_nrlssc_gnu_smoke_gx3_20x1_qc qc_te

**Comment by dabail10** [comment]
I downloaded from the PR tag (daveh150:update_plots).

**Comment by daveh150** [comment]
Hmm, that still looks like the basemap version. I'll email you what I committed so you can make sure it is the right version.

**Comment by dabail10** [comment]
There are two sets of plots here. The earlier ones were the old script.

**Comment by phil-blain** [comment]
That might be something on m side, but with the script in this PR I get plots like this;

image


I haven't yet looked into why it's all blue...

File: configuration/scripts/tests/QC/cice.t-test.py [hunk 1]
```diff
@@ -389,87 +389,200 @@ def plot_data(data, lat, lon, units, case, plot_type):
...
+        scSH = ax2.scatter(lon,lat,c=data,cmap=mycmap,s=4,edgecolors='none',
+                           vmin=cminSH, vmax=cmaxSH,
+                           transform=ccrs.PlateCarree())
 
+    else: 
         if plot_type == 'contour':
-            sc = m.contourf(x, y, d1, cmap='jet')
-        else:  # pcolor
-            sc = m.pcolor(x, y, d1, cmap='jet')
+            print("contour plot depreciated. using pcolor.")
```
Comment (phil-blain) [inline configuration/scripts/tests/QC/cice.t-test.py:473]:
```suggestion
            print("contour plot deprecated. using pcolor.")
```

**Comment by dabail10** [comment]
This all looks good to me. Not sure why Phillipe is getting blue everywhere. Is the colormap slightly different on your machine?

**Comment by phil-blain** [comment]
This is with Cartopy 0.20.2, Matpltlib 3.5.2. What versions are your using ?

**Comment by dabail10** [comment]
Cartopy 0.20.2 matplotlib 3.5.1

**Comment by phil-blain** [comment]
I can investigate on my own as to not delay merging this.

**Comment by phil-blain** [comment]
my plots are on the gx1 grid, maybe that's why I'm seeing something different?

**Comment by dabail10** [comment]
Just ran it on one of my gx1 tests and I don't see the blue.

ice_thickness_gx1testc
ice_thickness_gx1testb_minus_gx1testc
ice_thickness_gx1testb

**Comment by phil-blain** [comment]
In my plot I've just noticed there is also a dark longitudinal line near 30°E...

**Comment by phil-blain** [comment]
I've spent more time debugging this. I *think* it's due to the fact that the `t_lat` and `t_lon` arrays are masked arrays due to land proc elimination:

`t_lat`:
image
`t_lon`:
image

If I do something like [`t_lat=t_lat.filled(0)`](https://numpy.org/doc/stable/reference/maskedarray.generic.html#filling-in-the-missing-data), and the same for `t_lon`, i.e. filling the masked values with zeros:
image

Then I get the expected plot:

image

So @dabail10 I guess my next question would be what was your MPI x OpenMP configuration for the gx1 runs you used to test the new script ? Would it be possible that there was no land procs?

**Comment by phil-blain** [comment]
So I zoomed out a bit in my plot above and I don't think it's as easy as filling the lat/lon arrays with zeros:

image

we get spurious polygons, with different shapes depending on the filling value. I think we should use "complete" i.e. non masked `tlat`, `tlon`. This leads to the correct plot:

image


So one way to do it would be for the script to read the location of the input data directory (`CICE_data`) from the `env.machine_compiler` file  and then  read the grid from there to get the full lat/lon arrays, The only "problem" is that we currently do not have the gx1 grid as a NetCDF file.

**Comment by daveh150** [comment]
@phil-blain @dabail10 if the issue is related to masked locations in the tlon/tlat but NOT the data, could you try this to get the mask the same. Then I think the pcolor will ignore the masked locations:

data.mask = np.logical_or(data.mask,lon.mask)

**Comment by eclare108213** [comment]
In my opinion, tlat and tlon ought to be made 'complete' (no eliminated land blocks) in the output files.  One way to do this would be to write the original, global arrays, rather than writing them after the decomposition when the blocks are eliminated.  Not sure how easy this would be, or if it would work for PIO etc.

**Comment by daveh150** [comment]
Thanks @phil-blain. I think you had an idea to read the gx1 grid file instead of from the history file, if it was netcdf? That should be straight forward, we just need to write the TGRID in addition the UGRID.

**Comment by daveh150** [comment]
I made a grid_gx1.nc file. Hopefully it is attached, can y'all verify it is okay then perhaps add it to the CICE_data grid files for future use?
[grid_gx1.nc.gz](https://github.com/CICE-Consortium/CICE/files/9023675/grid_gx1.nc.gz)

Updating the cice.t-test.py to find this file will take a bit of thought, since it is not in the run directory.

**Comment by daveh150** [comment]
Sorry for quick modification of the gx1 netcdf, I changed the variables to be all lowercase in accordance with popgrid_nc. Also changed the kmt data type to integer and unit to 'none'.

[grid_gx1.nc.gz](https://github.com/CICE-Consortium/CICE/files/9023726/grid_gx1.nc.gz)

**Comment by phil-blain** [comment]
I implemented what we talked about in the previous meeting i.e. disabling land blocks elimination. 

@daveh150, you could update your branch to include [my commit](https://github.com/daveh150/CICE/compare/update_plots...phil-blain:CICE:update_plots) thusly:

```bash
git checkout update_plots
git pull git@github.com:phil-blain/CICE.git update_plots
git push
```

**Comment by apcraig** [comment]
Thanks for doing that @phil-blain, I think I volunteered but then dropped the ball.  @daveh150, could you pull @phil-blain's change to your sandbox and retest?  @phil-blain, have you confirmed that this fixes the issues you were seeing?  If so, and everyone agrees, we can problem merge soon.  Thanks!

**Comment by phil-blain** [comment]
I had already verified that using the full TLON/TLAT arrays made the script work correctly. I had done that by creating a gx1 test  with  `-p 1x1`, which then changed the block size and resulted in a single block and thus no land blocks elimination. I then manually loaded these TLON/TLAT arrays in Python and used those to create the "correct" maps in https://github.com/CICE-Consortium/CICE/pull/727#issuecomment-1170080421.

I just verified now that  with this updated `qc` option, i.e. including  ` distribution_wght = 'blockall'`, creating a test with `-p 44x1 -s qc` , i.e. [what the doc suggests using](https://cice-consortium-cice.readthedocs.io/en/master/user_guide/ug_testing.html#end-to-end-testing-procedure), the land blocks are indeed not eliminated and the TLON/TLAT arrays in the output files do not have any missing values. So I think this shows the fix is sufficient :)

**Comment by apcraig** [comment]
Thanks @phil-blain, it makes sense the blockall would fix the problem.  Thanks for confirming.

@daveh150, if there are any mods related to pulling in a separate lon/lat file for plotting, those can be removed again.

@daveh150, @phil-blain's mod is a one line change.  If you prefer to implement it manually rather than pull/push, that would be fine too.  You can see the mod here, https://github.com/daveh150/CICE/compare/update_plots...phil-blain:CICE:update_plots, as noted by @phil-blain.

**Comment by apcraig** [comment]
See #742

**Comment by daveh150** [comment]
I've been on leave since last week. I did not have any other changes for the lat/lon issue, glad that the blockall solution by @phil-blain worked! Thank you.

**Pull request closed without merging**